staff/lab/timecard

```diff
@@ -0,0 +1,139 @@
+    def adjusted_hour(hour):
+        """Converts a 0-23 hour format into a 1-12 hour format"""
+        if hour == 0 or hour == 12:
+            return 12
+        else:
+            return hour % 12
+
+    def pm_or_am(hour):
+        return 'am' if hour <= 11 else 'pm'
+
+    plt.switch_backend('agg')
+
+    hours = ['{}{}'.format(adjusted_hour(hour), pm_or_am(hour)) for hour in range(24)]
+    fig = plt.figure()
+    fig.set_size_inches(12, 4)
+    plt.xlim(0, 24)
+    plt.xticks(range(24), hours, fontsize=8)
+    plt.xlabel('Hour')
+    plt.ylabel('# Times Present')
+
+    # create a new mapping of user : length 24 array (frequencies for each hour)
+    users_daily_usage = dict()
+    for user, hours in usage.items():
+        users_daily_usage[user] = [sum(hours[h]) for h in range(24)]
+    stacked = [sum(x) for x in zip(*users_daily_usage.values())]
+    plt.ylim(0, max(stacked) * 1.15)  # set y limit to max cumulative usage
+
+    colors = ['b', 'y', 'g', 'r']
+    col_ind = 0
+    baseline = [0] * 24  # an offset on the bottom so stacked charts don't cover each other
+    for user in users:
+        hours = users_daily_usage[user]
+        plt.bar(
+            range(24),
+            hours,
+            align='edge',
+            color=colors[col_ind % len(colors)],
+            bottom=baseline,
+            label=user,
+        )
+        col_ind += 1
+        baseline = [x + y for x, y in zip(baseline, hours)]
+
+    plt.legend(loc='best', bbox_to_anchor=(1.13, 1.0), shadow=True, fancybox=True)
+    plt.title('OCF Lab Timecard for {} ({} to {})'.format(
+        ', '.join(users),
+        start.strftime('%Y-%m-%d'),
+        end.strftime('%Y-%m-%d')
+    ))
+    plt.savefig(dest, bbox_inches='tight')
+
+
+def get_usage_count(users, start=None, end=None):
+    sem_start, sem_end = stats.semester_dates()
+    if start is None:
+        start = datetime.datetime.combine(sem_start, datetime.time())
+    if end is None:
+        end = datetime.datetime.combine(sem_end, datetime.time())
+
+    step = datetime.timedelta(minutes=60)
+    hours = dict()
+
+    cnx = stats.get_connection()
+    cursor = cnx.cursor()
+
+    for user in users:
+        user_hours = collections.defaultdict(lambda: [0] * 7)
+        query = """
+            SELECT `start`, `end`
+            FROM `staff_session_duration_public`
+            WHERE `user` = %s
+                AND `start` >= %s
+                AND `end` <= %s
+                AND `duration` IS NOT NULL
+            ORDER BY start ASC"""
+        cursor.execute(query, (user, start, end))
+        logins = cursor.fetchall()
+
+        was_logged_in = dict()
+        for session in logins:
+            start_time = session['start'].replace(microsecond=0, second=0, minute=0)
+            end_time = session['end']
+            while start_time <= end_time:
+                was_logged_in[start_time] = True
+                start_time += step
+
+        time = start
+        while time <= end:
```

Comment:
I will bitch one last time about this has bad performance, especially when # of hours logged in << # of hours in semester (which is always in practice)

Reply:
I thought I would see if it was actually faster to flatten overlapping sessions instead of using @seanlobo's algorithm. So I created this patch: https://github.com/matthew-mcallister/utils/commit/aeb37a65ccb47cf8dec1922ae2a5b5fbe662fa78. Then I generated this graph:
![foo](https://user-images.githubusercontent.com/6722367/32467918-1591165e-c301-11e7-9d3e-696a20be0540.png)
Finally, I made the same performance measurement as @seanlobo:
```
Total query time: 0:00:01.150990
Total deduplication time: 0:00:00.000922
Total counting time: 0:00:00.004543
Total graphing time: 0:00:01.143586
```
Lo and behold, it's actually much slower than querying a dict once for each hour in the semester. That's Python for you!